ddl/delete_range.go

```diff
@@ -0,0 +1,315 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/ngaut/pools"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/terror"
+	"github.com/pingcap/tidb/util/sqlexec"
+)
+
+const (
+	insertDeleteRangeSQL   = `INSERT IGNORE INTO mysql.gc_delete_range VALUES ("%d", "%d", "%s", "%s", "%d")`
+	loadDeleteRangeSQL     = `SELECT job_id, element_id, start_key, end_key FROM mysql.gc_delete_range WHERE ts < %v ORDER BY ts`
+	completeDeleteRangeSQL = `DELETE FROM mysql.gc_delete_range WHERE job_id = %d AND element_id = %d`
+
+	delBatchSize = 65536
+	delBackLog   = 128
+)
+
+type delRangeManager interface {
+	addDelRangeJob(job *model.Job) error
+	start()
+	clear()
+}
+
+type delRange struct {
+	d            *ddl
+	ctxPool      *pools.ResourcePool
+	storeSupport bool
+	emulatorCh   chan struct{}
+	keys         []kv.Key
+}
+
+// newDelRangeManager returns a delRangeManager.
+func newDelRangeManager(d *ddl, ctxPool *pools.ResourcePool, supportDelRange bool) delRangeManager {
+	dr := &delRange{
+		d:            d,
+		ctxPool:      ctxPool,
+		storeSupport: supportDelRange,
+	}
+	if !dr.storeSupport {
+		dr.emulatorCh = make(chan struct{}, delBackLog)
+		dr.keys = make([]kv.Key, 0, delBatchSize)
+	}
+	return dr
+}
+
+// addDelRangeJob implements delRangeManager interface.
+func (dr *delRange) addDelRangeJob(job *model.Job) error {
+	resource, err := dr.ctxPool.Get()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer dr.ctxPool.Put(resource)
+	ctx := resource.(context.Context)
+	ctx.GetSessionVars().SetStatusFlag(mysql.ServerStatusAutocommit, true)
+
+	err = insertBgJobIntoDeleteRangeTable(ctx, job)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !dr.storeSupport {
+		dr.emulatorCh <- struct{}{}
+	}
+	log.Infof("[ddl] add job (%d,%s) into delete-range table", job.ID, job.Type.String())
+	return nil
+}
+
+// start implements delRangeManager interface.
+func (dr *delRange) start() {
+	if !dr.storeSupport {
+		dr.d.wait.Add(1)
+		go dr.startEmulator()
+	}
+}
+
+// clear implements delRangeManager interface.
+func (dr *delRange) clear() {
+	log.Infof("[ddl] closing delRange session pool")
+	dr.ctxPool.Close()
```

Comment:
yes, because `sysSessionPool` **factory** calling `createSession` to create a new session, which maybe causes some dead lock problems. so I use another pool with `createSessionWithDomain` to create session.